File: 295.数据流的中位数.py

from heapq import heappush, heappop


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
class MedianFinder:
    def __init__(self):
        self.maxpq = []
        self.minpq = []

    # store the num into maxpq
    # while keeping the size of maxpq >= size of minpq
    def addNum(self, num: int) -> None:
        heappush(self.maxpq, -num)
        heappush(self.minpq, -heappop(self.maxpq))
        if len(self.minpq) > len(self.maxpq):
            heappush(self.maxpq, -heappop(self.minpq))

    def findMedian(self) -> float:
        if len(self.maxpq) > len(self.minpq):
            return -self.maxpq[0]
        return (-self.maxpq[0] + self.minpq[0]) / 2

File: 295.数据流的中位数.cpp

#include <queue>
#include <vector>
#include <functional>

using std::priority_queue;
using std::vector;


/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
class MedianFinder {
public:
    // priority queue
    MedianFinder() {}

    // push the num into maxpq
    // keep maxpq larger than minpq by 1
    void addNum(int num) {
        maxpq.push(num);
        minpq.push(maxpq.top());
        maxpq.pop();
        if (minpq.size() > maxpq.size()) {
            maxpq.push(minpq.top());
            minpq.pop();
        }
    }

    double findMedian() {
        if (maxpq.size() > minpq.size()) {
            return maxpq.top();
        }
        return (maxpq.top() + minpq.top()) / 2.0;
    }

private:
    priority_queue<int, vector<int>, std::less<>> maxpq;
    priority_queue<int, vector<int>, std::greater<>> minpq;
};